cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/kafka/KafkaLogAppender.java

```
@@ -50,13 +60,36 @@
   @Override
   protected void append(LogMessage logMessage) {
     try {
-      byte [] bytes = loggingEventSerializer.toBytes(logMessage.getLoggingEvent(), logMessage.getLoggingContext());
-      producer.publish(logMessage.getLoggingContext().getLogPartition(), bytes);
+      String partitionKey = getPartitionKey(logMessage.getLoggingContext());
+      // publish only if its a valid type of partition key
+      if (!Strings.isNullOrEmpty(partitionKey)) {
+        byte [] bytes = loggingEventSerializer.toBytes(logMessage.getLoggingEvent(), logMessage.getLoggingContext());
+        producer.publish(partitionKey, bytes);
+      }
     } catch (Throwable t) {
       LOG.error("Got exception while serializing log event {}.", logMessage.getLoggingEvent(), t);
     }
   }
 
+  private String getPartitionKey(LoggingContext loggingContext) throws IllegalArgumentException {
+    String namespaceId = loggingContext.getSystemTagsMap().get(NamespaceLoggingContext.TAG_NAMESPACE_ID).getValue();
+
+    if (NamespaceId.SYSTEM.getNamespace().equals(namespaceId)) {
+      return loggingContext.getLogPartition();
+    }
+
+    switch (LogPartitionType.valueOf(cConf.get(Constants.Logging.LOG_PUBLISH_PARTITION_KEY).toUpperCase())) {
+      case PROGRAM:
+        return loggingContext.getLogPartition();
+      case APPLICATION:
+        return namespaceId + ":" + loggingContext.getSystemTagsMap().get(ApplicationLoggingContext
+                                                                           .TAG_APPLICATION_ID).getValue();
+      default:
+        // this should never happen
+        return null;
```

Should throw instead of returning `null`.